x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/dataframe/analyses/Classification.java

```diff
@@ -406,8 +405,38 @@ public boolean supportsCategoricalFields() {
         return additionalProperties;
     }
 
-    private static Object extractMapping(String path, Map<String, Object> mappingsProperties) {
-        return extractValue(String.join(".properties.", path.split("\\.")), mappingsProperties);
+    // Visible for testing
+    static Object extractMapping(String path, Map<String, Object> mappingsProperties) {
+        if (path.isEmpty()) {
+            return mappingsProperties;
+        }
+        Object current = mappingsProperties;
+        Map<String, Object> currentAsMap = mappingsProperties;
+        String[] pathParts = path.split("\\.");
+        for (int i = 0; i < pathParts.length; ++i) {
+            if (current == null) {
+                return null;
+            }
+            String pathPart = pathParts[i];
+            if (currentAsMap.containsKey(pathPart) == false) {
+                return null;
+            }
+            current = currentAsMap.get(pathPart);
+            if (i == pathParts.length - 1) {
+                break;
+            }
+            currentAsMap = (Map<String, Object>) current;
+            if (currentAsMap.containsKey("properties")) {
+                current = currentAsMap.get("properties");
+                currentAsMap = (Map<String, Object>) current;
+            } else if (currentAsMap.containsKey("fields")) {
+                current = currentAsMap.get("fields");
+                currentAsMap = (Map<String, Object>) current;
+            } else {
+                return null;
+            }
```

Comment:
```suggestion
            currentAsMap = (Map<String, Object>) current;
            current = currentAsMap.get("properties");
            if (current == null) {
                current = currentAsMap.get("fields");
            }
            currentAsMap = (Map<String, Object>) current;
```
You do a `null` check at the start of the loop. So, use that as your loop exit. Seems like only one branch is needed here.

Reply:
Also, casting `null` ends up as `null`. So that should still be ok.